src/download.rs

```diff
@@ -0,0 +1,300 @@
+///
+/// Some messages as non-delivery-reports (NDN) or read-receipts (MDN)
+/// need to be downloaded completely to handle them correctly,
+/// eg. to assign them to the correct chat.
+/// As these messages are typically small,
+/// they're catched by `MIN_DOWNLOAD_LIMIT`.
+const MIN_DOWNLOAD_LIMIT: u32 = 32768;
+
+/// If `delete_server_after` is set to small timeouts (eg. "at once"),
+/// the user might have no chance to actually download a message.
+/// `MIN_DELETE_SERVER_AFTER` increases the timeout in this case.
+pub const MIN_DELETE_SERVER_AFTER: i64 = 48 * 60 * 60;
+
+#[derive(
+    Debug,
+    Display,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    FromPrimitive,
+    ToPrimitive,
+    FromSql,
+    ToSql,
+    Serialize,
+    Deserialize,
+)]
+#[repr(u32)]
+pub enum DownloadState {
+    Done = 0,
+    Available = 10,
+    Failure = 20,
+    InProgress = 1000,
+}
+
+impl Default for DownloadState {
+    fn default() -> Self {
+        DownloadState::Done
+    }
+}
+
+impl Context {
+    // Gets validated download limit or 0 for "no limit".
+    pub(crate) async fn get_download_limit(&self) -> Result<u32> {
+        let download_limit = self.get_config_int(Config::DownloadLimit).await?;
+        if download_limit <= 0 {
+            Ok(0)
+        } else {
+            Ok(max(MIN_DOWNLOAD_LIMIT, download_limit as u32))
+        }
+    }
+}
+
+impl MsgId {
+    /// Adds the job `Action::DownloadMsg` to download a message.
+    pub async fn download_full(self, context: &Context) -> Result<()> {
+        let msg = Message::load_from_db(context, self).await?;
+        match msg.get_download_state() {
+            DownloadState::Done => return Err(anyhow!("Nothing to download.")),
+            DownloadState::InProgress => return Err(anyhow!("Download already in progress.")),
+            DownloadState::Available | DownloadState::Failure => {
+                self.update_download_state(context, DownloadState::InProgress)
+                    .await?;
+                job::add(
+                    context,
+                    Job::new(Action::DownloadMsg, self.to_u32(), Params::new(), 0),
+                )
+                .await;
+            }
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn update_download_state(
+        self,
+        context: &Context,
+        download_state: DownloadState,
+    ) -> Result<()> {
+        let msg = Message::load_from_db(context, self).await?;
+        context
+            .sql
+            .execute(
+                "UPDATE msgs SET download_state=? WHERE id=?;",
+                paramsv![download_state, self],
+            )
+            .await?;
+        context.emit_event(EventType::MsgsChanged {
+            chat_id: msg.chat_id,
+            msg_id: self,
+        });
+        Ok(())
+    }
+}
+
+impl Message {
+    /// Gets the download state of the message.
+    pub fn get_download_state(&self) -> DownloadState {
```

Comment:
Normally getters in Rust don't use `get_` prefix: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

I'd name it `download_state` and change doc-string to "Returns ...".

Reply:
k, although we're using `get_` at other places, it maybe makes sense to follow rust-conventions for new additions. i also rustified `get_download_limit()` to `download_limit()`.